src/command.rs

```diff
@@ -49,6 +51,95 @@ macro_rules! delegate {
     }};
 }
 
+/// If a command is `OverSsh` then it can be executed over an SSH session.
+/// Primarily a way to allow `std::process::Command` to be turned directly into an `openssh::Command`.
+pub trait OverSsh {
+    /// Given a session, return a command that can be executed over that session.
+    ///
+    /// **Note**: The command to be executed on the remote machine does not include
+    /// any environment variables or the current directory. They must be
+    /// set explicitly, if desired.
+    ///
+    /// # Example
+    /// Consider the implementation of `OverSsh` for `std::process::Command`:
+    ///
+    /// ```no_run
+    /// # #[tokio::main(flavor = "current_thread")]
+    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    ///     use std::process::Command;
+    ///     use openssh::{Session, KnownHosts, OverSsh};
+    ///
+    ///     let session = Session::connect_mux("me@ssh.example.com", KnownHosts::Strict).await?;
+    ///     let ls =
+    ///         Command::new("ls")
+    ///         .arg("-l")
+    ///         .arg("-a")
+    ///         .arg("-h")
+    ///         .over_session(&session)
+    ///         .output()
+    ///         .await?;
+    ///
+    ///     assert!(String::from_utf8(ls.stdout).unwrap().contains("total"));
+    /// #   Ok(())
+    /// }
+    ///
+    /// ```
+    fn over_session<'session>(&self, session: &'session Session) -> crate::Command<'session>;
+}
+
+impl OverSsh for std::process::Command {
+    fn over_session<'session>(&self, session: &'session Session) -> Command<'session> {
+        let program_escaped: Cow<'_, OsStr> = escape(self.get_program());
+        let mut command = session.raw_command(program_escaped);
+
+        let args = self.get_args().map(escape);
+        command.raw_args(args);
+        command
+    }
+}
+
+impl OverSsh for tokio::process::Command {
+    fn over_session<'session>(&self, session: &'session Session) -> Command<'session> {
+        self.as_std().over_session(session)
+    }
+}
+
+impl<S> OverSsh for &S
+where
+    S: OverSsh,
+{
+    fn over_session<'session>(&self, session: &'session Session) -> Command<'session> {
+        <S as OverSsh>::over_session(self, session)
+    }
+}
+
+impl<S> OverSsh for Box<S>
+where
+    S: OverSsh,
+{
+    fn over_session<'session>(&self, session: &'session Session) -> Command<'session> {
+        <S as OverSsh>::over_session(self, session)
+    }
+}
+
+impl<S> OverSsh for std::rc::Rc<S>
+where
+    S: OverSsh,
+{
+    fn over_session<'session>(&self, session: &'session Session) -> Command<'session> {
+        <S as OverSsh>::over_session(self, session)
+    }
+}
+
+impl<S> OverSsh for std::sync::Arc<S>
+where
+    S: OverSsh,
+{
+    fn over_session<'session>(&self, session: &'session Session) -> Command<'session> {
+        <S as OverSsh>::over_session(self, session)
+    }
+}
+
```

Comment:
I think these probably aren't worth including. Even the impl for `Box` feels a little odd. I also think `impl for &mut` is missing, if anything (but again, I think just having the `impl for &` is probably good enough).

Reply:
I think impl for &T can also handle &mut T given rust auto-deref rules?

Reply:
I think that's only sort of true. In particular it still means that `&mut OverSssh` doesn't implement `OpenSsh`, which can be annoying in certain cases.